src/kontrol/foundry.py

```diff
@@ -155,14 +157,145 @@ def _exec_rename_custom_step(self, cterm: CTerm) -> KCFGExtendResult | None:
         _LOGGER.info(f'Renaming {target_var.name} to {name}')
         return Step(CTerm(new_cterm.config, constraints), 1, (), ['foundry_rename'], cut=True)
 
-    def custom_step(self, cterm: CTerm, _cterm_symbolic: CTermSymbolic) -> KCFGExtendResult | None:
+    def _check_forget_pattern(self, cterm: CTerm) -> bool:
+        """Given a CTerm, check if the rule 'FOUNDRY-ACCOUNTS.forget' is at the top of the K_CELL.
+        This method checks if the 'FOUNDRY-ACCOUNTS.forget' rule is at the top of the `K_CELL` in the given `cterm`.
+        If the rule matches, the resulting substitution is cached in `_cached_subst` for later use in `custom_step`
+        :param cterm: The CTerm representing the current state of the proof node.
+        :return: `True` if the pattern matches and a custom step can be made; `False` otherwise.
+        """
+        abstract_pattern = KSequence(
+            [
+                KApply('cheatcode_forget', [KVariable('###TERM1'), KVariable('###OPERATOR'), KVariable('###TERM2')]),
+                KVariable('###CONTINUATION'),
+            ]
+        )
+        self._cached_subst = abstract_pattern.match(cterm.cell('K_CELL'))
+        return self._cached_subst is not None
```

Comment:
The optimization of caching the `_cached_subst` seems pretty unnecessary to me. As far as I can tell, it will only fire _if_ `_check_forget_pattern` is true, which should only happpen when we have `cheatcode_forget` at the top of the K cell, which only happens maybe once or twice a test? So we may set the cached result once, and use it once, but no more than that, in a given test. So it introduces state dependencies/synchronization between `can_take_custom_step` and `custom_step` for maybe less than a second of savings a single time in a proof.

Reply:
Do we have data about how much time it's saving?

Reply:
We can change `_check_forget_pattern` to return a `Subst | None`, and it can still be tested as a boolean where it's used.